File: 2022/2022W15/preppindataW15.R
library(readxl)
library(dplyr)
library(lubridate)

today <- as.POSIXct('2022-04-13', tz ='UTC')

final <- read_excel("C:/Data/PreppinData/Rental Contracts.xlsx") %>% 
  mutate('Contract Length' = round(time_length(floor_date(`Contract End`,'month') - floor_date(`Contract Start`,'month'), 'month')),
         'Months Until Expiry' =  round(time_length(floor_date(`Contract End`,'month') - floor_date(today,'month'), 'month'))) %>%
  merge(read_excel("C:/Data/PreppinData/Office Space Prices.xlsx"), by=c('City', 'Office Size')) %>%
  .[rep(seq_len(nrow(.)), .[['Contract Length']]),] %>%
  group_by(`ID`) %>%
  mutate('Seq' = row_number(`ID`),
         'Month Divider' = `Contract Start` %m+% months(`Seq`-1),
         'Cumulative Monthly Cost' = `Rent per Month`*`Seq`) 
  
output1 <- final %>%
  select(c('Cumulative Monthly Cost', 'ID', 'Country', 'City', 'Address', 'Company', 'Office Size',
           'Contract Start', 'Contract End', 'Contract Length', 'Months Until Expiry', 'People', 'Per Person', 
           'Rent per Month', 'Month Divider'))
  
output2 <- final %>%
  mutate('Year' = year(`Month Divider`),
         'Paid Rent' = if_else(floor_date(`Month Divider`,'month')<=floor_date(today,'month'), `Rent per Month`, 
                               if_else(year(`Month Divider`)<=year(today), 0, NA_real_))) %>%
  group_by(`Year`) %>%
  summarise('EoY and Current' = sum(`Paid Rent`))
  
View(output1)
View(output2)

File: 2022/2022W15/PreppindataW15.py
import pandas as pd
import datetime
import numpy as np
from dateutil.relativedelta import relativedelta

today = datetime.date(2022, 4, 13)

contract = pd.read_excel(pd.ExcelFile(r"C:\Data\PreppinData\Rental Contracts.xlsx"))
contract['Contract Length'] = contract.apply(lambda x: (x['Contract End'].year - x['Contract Start'].year)*12 + 
                                             (x['Contract End'].month - x['Contract Start'].month), axis=1)
contract['Months Until Expiry'] = contract.apply(lambda x: (x['Contract End'].year - today.year)*12 + 
                                             (x['Contract End'].month - today.month), axis=1)

price = pd.read_excel(pd.ExcelFile(r"C:\Data\PreppinData\Office Space Prices.xlsx"))

final = pd.merge(contract, price, on=['City', 'Office Size'])
final = final.iloc[np.repeat(np.arange(len(final)), list(final['Contract Length']))].reset_index(drop=True)
final['Seq'] = final.groupby(['ID'])['ID'].rank(method='first').astype(int)
final['Month Divider'] = final.apply(lambda x: x['Contract Start'] + relativedelta(months=x['Seq']-1),axis=1) 
final['Cumulative Monthly Cost'] = final['Rent per Month']*final['Seq'] 

output1 = final[['Cumulative Monthly Cost', 'ID', 'Country', 'City', 'Address', 'Company', 'Office Size',
       'Contract Start', 'Contract End', 'Contract Length', 'Months Until Expiry', 'People', 'Per Person', 
       'Rent per Month', 'Month Divider']]

final['Year'] = final['Month Divider'].apply(lambda x: x.year)
final['EoY and Current'] = final.apply(lambda x: x['Rent per Month'] if x['Month Divider'].date().replace(day=1)<=today.replace(day=1) else 0, axis=1)

output2 = final.groupby('Year', as_index=False).agg({'EoY and Current':'sum'})
output2['EoY and Current'] = output2.apply(lambda x: x['EoY and Current'] if x['Year']<=today.year else None, axis=1)
